Django setup guide using cookiecutter (#48)
rootstrap/tech-guides: python/cookiecutter-django.md

> # How starting a new Django project with Cookiecutter

Reviewer: ```suggestion
# How to start a new Django project with Cookiecutter
```

> ## Install Cookiecutter

Reviewer 2: Here I would explain very shortly what does cookiecutter do. And also a link to cookiecutter

`$ pip install cookiecutter`

## Create a new project
`$ cookiecutter gh:pydanny/cookiecutter-django`

## Fill generator options
*Cookiecutter* gives us a set of pre-defined configurations to include at our project. Each configuration depends on each project that you will be involved in. You can take a look at [the full list of generation options](https://cookiecutter-django.readthedocs.io/en/latest/project-generation-options.html).

> Despite the previously mentioned, we provide you a list of the principal configuration that you have to configure in a new *Cookiecutter* project.

Reviewer 2: ```suggestion
Despite the previously mentioned, we provide you a list of the principal configuration that you have to set in a new *Cookiecutter* project.
```

WDYT? To avoid being repetitive

- **use_pycharm**: This option asks us if we want to add some configuration files to run the project at PyCharm, the Jetbrains' Python IDE. It's personal, choose `n` or `y` according to your preferences.
- **cloud_provider**: Choose the cloud provider for files resources that will be attached from the models. At Rootstrap we use AWS, so, you should select option `1`.
- **mail_service**: Choose the email service for send emails from the application. At Rootstrap we use Sendgrid, so, you should select option `6`.
- **use_drf**: This option asks us if we want to add [Django REST framework](https://www.django-rest-framework.org/) to our application. At Rootstrap we use it as the base of our API, so, you should enter the option: `y`.
- **use_mailhog**: This option asks us if we want to add [Mailhog](https://github.com/mailhog/MailHog) as an email viewer during the development process. We recommend that option, so, you should enter: `y`.
- **use_heroku**: This option asks us if we want to add some configuration files to deploy the project to *Heroku*. At Rootstrap we usually deploy to it, so, you should enter: `y`.

> - **keep_local_envs_in_vcs**: This option asks us if we want to add the environment configurations to the Version Control System. Also, it's useful to set Heroku or docker variables. AAt Rootstrap we manage that information differently, so, you should enter: `n`.

Reviewer 2: Typo here: `AAt Rootstrap we...`

author_name [Daniel Roy Greenfeld]: Rootstrap, Inc.
domain_name [example.com]: example.com
email [rootstrap@example.com]: info@example.com
version [0.1.0]:
Select open_source_license:
1 - MIT
2 - BSD
3 - GPLv3
4 - Apache Software License 2.0
5 - Not open source
Choose from 1, 2, 3, 4, 5 [1]: 1
timezone [UTC]:
windows [n]:
use_pycharm [n]:
use_docker [n]:
Select postgresql_version:
1 - 12.3
2 - 11.8
3 - 10.8
4 - 9.6
5 - 9.5
Choose from 1, 2, 3, 4, 5 [1]: 1
Select js_task_runner:
1 - None
2 - Gulp
Choose from 1, 2 [1]:
Select cloud_provider:
1 - AWS
2 - GCP
3 - None
Choose from 1, 2, 3 [1]: 1
Select mail_service:
1 - Mailgun
2 - Amazon SES
3 - Mailjet
4 - Mandrill
5 - Postmark
6 - Sendgrid
7 - SendinBlue
8 - SparkPost
9 - Other SMTP
Choose from 1, 2, 3, 4, 5, 6, 7, 8, 9 [1]: 6
use_async [n]:
use_drf [n]: y
custom_bootstrap_compilation [n]: n
use_compressor [n]:
use_celery [n]:
use_mailhog [n]: y
use_sentry [n]:
use_whitenoise [n]:
use_heroku [n]: y
Select ci_tool:
1 - None
2 - Travis
3 - Gitlab
Choose from 1, 2, 3 [1]:
keep_local_envs_in_vcs [y]: n
debug [n]:
```

</details>

## Install pipenv

> The recommendation from Rootstrap is that you use `pipenv` to manage the development environment dependencies.

Reviewer 2: Here I would add a link to pipenv

To install the package, just run `$ pip install pipenv`.

## Install dependencies
To install the dev dependencies, run `$ pipenv install -r requirements/local.txt --dev`.

NOTE:
- Remember, when you install a new dependency, check if it will be needed at production or not. In the case that it won't be needed, add the flag `--dev` at the end of the installation command.

> - At pipenv to run a command, you must write before `pipenv run`, for example, the command that runs the server should be `$ pipenv run python manage.py runserver`.

Reviewer 2: Also you can add that you can execute pipenv shell in the folder to "enter to the environment" and there you don't need to execute `pipenv run` before ` python manage.py runserver`.

Author: Sure! I didn't know that trick. I'll add it

```

## Migrate
`$ pipenv run python manage.py migrate`

## Run test
`$ pipenv run pytest`

## Run test and get coverage percentage
`$ pipenv run coverage run -m pytest && pipenv run coverage report -m`

NOTE:
- The command must run without exceptions because it mandatory for future configurations.
- Don't matter if you write tests following the [unittest](https://docs.python.org/3/library/unittest.html) or [pytest](https://docs.pytest.org/) approach since `$ pipenv run pytest` command will be able to run both of them.

## Install Mailhog
`$ brew install mailhog`

NOTE: If you prefer other options to install *Mailhog*, you can take a look at the [Mailhog installation guideline](https://github.com/mailhog/MailHog#installation).

Start the server running `$ mailhog`

Finally, if you go to `http://localhost:8025/`, your mail server is running.

## Code Style
### [flake8](https://flake8.pycqa.org/) and [flake8-isort](https://github.com/gforcada/flake8-isort)
*NOTE: By default this package is already installed.*

1. `$ pipenv install flake8 flake8-isort --dev`
1. Add dependencies to requirement files. Below the *Code quality* separator:
```
flake8==<version>
flake8-isort==<version>
```

> 3. Add custom configuration at `setup.cfg` file according to the Rootstrap's standards.

Reviewer 2: Can we add here a link on `Rootstrap's standards`? or at least a file for example?

    - **extend-ignore**: exclude flake8 warnings to avoid black conflicts. [See more.](https://black.readthedocs.io/en/stable/compatible_configs.html#id2)
```yml
#  setup.cfg

[flake8]
max-line-length = 120
exclude = .tox,.git,*/migrations/*,*/static/CACHE/*,docs,node_modules,venv/*
extend-ignore = E203, W503
```
4. Check running `$ pipenv run flake8`.
    - This command list all the issues found.
    - A useful command could be `$ pipenv run isort .` which resolves all the issues related to import styles. Also, you can add the flag `--interactive` at the end of the command to resolve or ignore one issue by one.

### [black](https://black.readthedocs.io/)
*NOTE: By default this package is already installed.*

1. `$ pipenv install black --dev`
1. Add dependencies to requirement files. Below the *Code quality* separator:
```
black==<version>
```

> 3. Add custom configuration at `pyproject.toml` file according to the Rootstrap's standards.

Reviewer 2: Can we add here a link on Rootstrap's standards? or at least a file for example?

Reviewer: I think it is being done in line 197

Reviewer 2: Oh my bad I thought that we need more configuration lines and that was an example

        restore-keys: |
          ${{ runner.os }}-venv-

    - name: Installing requirements
      run: |
        pipenv install --dev

    - name: Checking PEP8 code style
      run: |
        pipenv run flake8 --count

    - name: Checking Black code formatter
      run: |
        pipenv run black . --check

    - name: Running tests
      run: |
        pipenv run coverage run -m pytest --ds=config.settings.test

    - name: Checking coverage
      run: |
        pipenv run coverage report --fail-under=90 -m
        pipenv run coverage xml

    - name: Setup Code Climate test-reporter
      run: |
          curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter
          chmod +x ./cc-test-reporter
          ./cc-test-reporter before-build
          ./cc-test-reporter after-build --coverage-input-type coverage.py --exit-code $?
```

</details>

## Adding a new app

1. To create a new app into the project, we have to run the command `$ pipenv run python manage.py startapp <new_app_name>`.
    - **new_app_name**: is the name of the new app and it has to be written in plural. For example: targets

> 1. How that command creates the new app at the root directory, you have to move it to `<project_name>` directory, where there are other existing apps like `users`.

Reviewer 2: ```suggestion
1. Since that command creates the new app at the root directory, you have to move it to `<project_name>` directory, where there are other existing apps like `users`.
```

# How to start a new Django project with Cookiecutter
The aim of this guide is to help you to initialize your Django project from one of the most recognized open-source utilities for this area.

## Install Cookiecutter
[*Cookiecutter*](https://github.com/cookiecutter/cookiecutter) is a command-line tool that helps us to create projects from different templates, in our case, we will create a Django project.

Firstly, we have to install that tool running `$ pip install cookiecutter`

## Create a new project

> How we want to create a Django project, we need to use the [Cookiecutter Django](https://github.com/pydanny/cookiecutter-django) template.

Reviewer 2: ```suggestion
Since we want to create a Django project, we need to use the [Cookiecutter Django](https://github.com/pydanny/cookiecutter-django) template.
```
